projects/common/src/time/user-specified-time-range.service.ts

```diff
@@ -0,0 +1,76 @@
+import { Injectable } from '@angular/core';
+import { isNil } from 'lodash-es';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { distinctUntilChanged, map, share } from 'rxjs/operators';
+import { NavigationService } from '../navigation/navigation.service';
+import { PreferenceService, StorageType } from '../preference/preference.service';
+import { TimeRange } from './time-range';
+import { TimeRangeService } from './time-range.service';
+
+@Injectable({ providedIn: 'root' })
+export class UserSpecifiedTimeRangeService {
+  private static readonly STORAGE_TYPE: StorageType = StorageType.Session;
+  private static readonly TIME_RANGE_PREFERENCE_KEY: string = 'page-time-range';
+
+  private readonly pageTimeRangesSubject$: BehaviorSubject<PageTimeRangeMap> = new BehaviorSubject<PageTimeRangeMap>(
```

Comment:
yeah, this is something we wanted to clean up. The purpose of maintaining `pageTimeRangesSubject` is to get the last value for the `set` operation.

We can use the `tap` operator to store the last emitted value. Here is an example
```
import { Injectable } from '@angular/core';
import { isNil } from 'lodash-es';
import { Observable } from 'rxjs';
import { distinctUntilChanged, map, share, tap } from 'rxjs/operators';
import { NavigationService } from '../navigation/navigation.service';
import { PreferenceService, StorageType } from '../preference/preference.service';
import { TimeRange } from './time-range';
import { TimeRangeService } from './time-range.service';

@Injectable({ providedIn: 'root' })
export class UserSpecifiedTimeRangeService {
  private static readonly STORAGE_TYPE: StorageType = StorageType.Session;
  private static readonly TIME_RANGE_PREFERENCE_KEY: string = 'page-time-range';

  private readonly pageTimeRanges$: Observable<PageTimeRangeMap>;
  private currentTimeRangeMap?: PageTimeRangeMap;

  public constructor(
    private readonly preferenceService: PreferenceService,
    private readonly timeRangeService: TimeRangeService,
    private readonly navigationService: NavigationService
  ) {
    this.pageTimeRanges$ = this.buildPageTimeRangeObservable();
  }

  public getUserSpecifiedTimeRangeForPage(path: string): Observable<TimeRange> {
    return this.pageTimeRanges$.pipe(
      distinctUntilChanged((prev, curr) => prev[path] === curr[path]),
      map(timeRanges => {
        if (isNil(timeRanges[path])) {
          const timeRangeForPath = this.getDefaultPageTimeRange(path);
          this.setUserSpecifiedTimeRangeForPage(path, timeRangeForPath);

          return timeRangeForPath;
        }

        return this.timeRangeService.timeRangeFromUrlString(timeRanges[path]);
      }),
      share()
    );
  }

  public setUserSpecifiedTimeRangeForPage(path: string, value: TimeRange): void {
    const newMap: PageTimeRangeMap = { ...(this.currentTimeRangeMap ?? {}), [path]: value.toUrlString() };
    this.preferenceService.set(
      UserSpecifiedTimeRangeService.TIME_RANGE_PREFERENCE_KEY,
      newMap,
      UserSpecifiedTimeRangeService.STORAGE_TYPE
    );
  }

  private buildPageTimeRangeObservable(): Observable<PageTimeRangeMap> {
    return this.preferenceService
      .get<PageTimeRangeMap>(
        UserSpecifiedTimeRangeService.TIME_RANGE_PREFERENCE_KEY,
        {},
        UserSpecifiedTimeRangeService.STORAGE_TYPE
      )
      .pipe(tap(timeRangeMap => (this.currentTimeRangeMap = timeRangeMap)));
  }

  private getDefaultPageTimeRange(path: string): TimeRange {
    const defaultTimeRange = this.navigationService.getRouteConfig([path], this.navigationService.rootRoute())?.data
      ?.defaultTimeRange;

    if (!defaultTimeRange) {
      //  Use current time range as default default
      return this.timeRangeService.getCurrentTimeRange();
    }

    return defaultTimeRange;
  }
}

interface PageTimeRangeMap {
  [path: string]: string;
}

```

Reply:
Yeah bad name on L15, it should be `storedTimeRangesSubject$`. It's purpose is to provide the set method async access to the values that are currently stored (L62).  Because the values are stored in one object in storage, the set method requires access to the currently stored values before setting/updating a new one

Reply:
Is this to avoid calling preferenceService.get more than once? If so, that doesn't save us any perf and isn't worth any complexity (go look at preference service, it's not a server call or anything, it's an observable to support watching for changes).

Reply:
> is what to avoid calling preferenceService.get more that once, the single object structure ?

Nope, I meant the mixture of sync and async. I would encourage just having get/set in this service be wrappers around the get/set in the preference service. In the set case, just call get first, then merge the new value into the set call once we receive the get result.